odl/operator/fspace_ops.py

```diff
@@ -0,0 +1,412 @@
+            out[slice1] = (3.0 * f_data[slice2] - 4.0 * f_data[slice3] +
+                           f_data[slice4]) / 2.0
+
+    # divide by step size
+    out /= dx
+
+    return out
+
+
+# noinspection PyAbstractClass
+class DiscreteGradient(Operator):
+    """Gradient operator for any number of dimensions `n`.
+
+    Calls helper function `discrete_part_deriv` to calculate each component
+    of the resulting product space vector. For the adjoint of the `Gradient`
+    operator to match the negative `DiscreteDivergence` operator requires
+    `zero_padding`.
+    """
+
+    def __init__(self, space, voxel_size=(1,), edge_order=2,
+                 zero_padding=True):
+        """Initialize a `DiscreteGradient` operator instance.
+
+        Zero padding is required for the negative `DiscreteDivergence`
+        operator to match the adjoint of the `DiscreteGradient` operator.
+
+        Parameters
+        ----------
+        space : `FunctionSpace`
+            The space of elements which the operator is acting on
+        voxel_size : `n`-tuple of `floats`
+            `n`-tuple of scalars specifying the sample distances in each
+            dimension. Default distance is 1 in each direction
+        edge_order : {1, 2}, optional
+            Partial derivative is calculated using Nth order accurate
+            differences at the boundaries. Default edge order: 2
+        zero_padding : `bool`, optional
+            Implicit zero padding. Assumes values outside of `space` to be
+            zero. Default: True
+        """
+        self.voxel_size = voxel_size
+        self.edge_order = edge_order
+        self.zero_padding = zero_padding
+        super().__init__(domain=space,
+                         range=ProductSpace(space, len(voxel_size)),
+                         linear=True)
+
+    def _apply(self, rhs, out):
+        """Apply gradient operator to `rhs` and store result in `out`.
+
+        Parameters
+        ----------
+        rhs : ``domain`` element
+            Input vector to which the `DiscreteGradient` operator is applied to
+        out : ``range`` element
+            Output vector to which the result is written
+
+        Examples
+        --------
+        >>> from odl import uniform_discr, FunctionSpace, IntervalProd
+        >>> def ndvolume(vol_size, ndim, dtype=None):
+        ...     s = [1]
+        ...     vol = np.arange(vol_size, dtype=dtype)
+        ...     for _ in range(ndim - 1):
+        ...         s.insert(0, vol_size)
+        ...         vol = vol * vol.reshape(s)
+        ...     return vol
+        >>> ndim = 3
+        >>> vsize = 100
+        >>> disc = uniform_discr(FunctionSpace(IntervalProd(
+        ... [0.]*ndim, [vsize]*ndim)), [vsize]*ndim)
+        >>> f = disc.element(ndvolume(vsize, ndim, np.int32))
+        >>> A = DiscreteGradient(disc, (1.,)*ndim, zero_padding=True)
+        >>> Af = A(f)
+        >>> g = A.range.one()
```

Comment:
No, but you take the adjoint of a one, which should be zero?

Id prefer a example in one dimension that is very simple to follow. These types of examples should go into the test suite.